### PR TITLE
feat(wasm): bump the spin shim the latest version from SpinKube

### PR DESCRIPTION
**Pull request by Mossaka**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

Unlike #4695 , this PR does not remove the existing Wasm shims, including `Slight` and `WWS`. Instead, it only bumps the spin shim to the latest version `v0.15.1` released from the SpinKube organization (instead of Deislabs). It will replace #4695 

**Which issue(s) this PR fixes**:

Fixes #

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:

```
none
```

### CODE REVIEW

**Comment by Mossaka** [comment]
@cameronmeissner Could you please take a look at this PR? 🙏

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (cameronmeissner) [inline parts/linux/cloud-init/artifacts/cse_install.sh:148]: we've been consolidating component declaration towards components.json: https://github.com/Azure/AgentBaker/blob/master/parts/linux/cloud-init/artifacts/components.json, probably a good idea to move this over there as well

File: nbcparser/pkg/parser/templates/containerd.toml.gtpl
Comment (cameronmeissner) [inline nbcparser/pkg/parser/templates/containerd.toml.gtpl:48]: I'm assuming this is updating the "default"/"latest" spin version? will updating this still allow v0-3-0 to be used as specified by RP?

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (cameronmeissner) [reply parts/linux/cloud-init/artifacts/cse_install.sh:148]: on second thought, we can probably do that in a separate PR :)

Comment (cameronmeissner) [reply parts/linux/cloud-init/artifacts/cse_install.sh:148]: cc @Devinwong

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (cameronmeissner) [inline parts/linux/cloud-init/artifacts/cse_install.sh:142]: I'm realizing now that we don't have any content tests around installed wasm shims, probably a good idea to add a new test here to make sure all this install logic is working as expected - content tests are ran at VHD build time: https://github.com/Azure/AgentBaker/blob/master/vhdbuilder/packer/test/linux-vhd-content-test.sh

File: nbcparser/pkg/parser/templates/containerd.toml.gtpl
Comment (devigned) [reply nbcparser/pkg/parser/templates/containerd.toml.gtpl:48]: Yes. If the `spin` runtime is used, it will use the latest version. If the `spin-v0-3-0` is used, then it would use v0-3-o.

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (devigned) [reply parts/linux/cloud-init/artifacts/cse_install.sh:142]: Will do!

**Comment by devigned** [comment]
@cameronmeissner I don't think I have access to update this branch, and @Mossaka is out on holiday. I have authored a test which you can see here: https://github.com/Azure/AgentBaker/compare/jiazho/bump-spinkube...devigned:bump-spinkube?expand=1.

I'm happy to provide the test as a follow on PR if we were able to get this PR through. If this PR isn't going to make it, I'm happy to open a new PR with this PR's content + the test. Which would you prefer?

**Comment by Mossaka** [comment]
@cameronmeissner I noticed that the AgentBaker E2E tests failed for Wasm in the CI. I tried to run the e2e tests on local, but I got 

```
--- FAIL: Test_azurelinuxv2 (4.05s)
panic: failed to get RG "abe2e-westus3": HEAD https://management.azure.com/subscriptions/8ecadfc9-d1a3-4ea4-b844-0d9f87e4d7c8/resourcegroups/abe2e-westus3
        --------------------------------------------------------------------------------
        RESPONSE 403: 403 Forbidden
        ERROR CODE UNAVAILABLE
        --------------------------------------------------------------------------------
        Response contained no body
        --------------------------------------------------------------------------------
         [recovered]
        panic: failed to get RG "abe2e-westus3": HEAD https://management.azure.com/subscriptions/8ecadfc9-d1a3-4ea4-b844-0d9f87e4d7c8/resourcegroups/abe2e-westus3
        --------------------------------------------------------------------------------
        RESPONSE 403: 403 Forbidden
        ERROR CODE UNAVAILABLE
        --------------------------------------------------------------------------------
        Response contained no body
        --------------------------------------------------------------------------------
```

I think that's because my azure account does not have the necessary permissions. Could there be something I didn't set up on my devbox properly? Do you have any instructions on how to run E2E tests locally, including setting up the environment and authenticate?

**Comment by Mossaka** [comment]
I might have missed something but where can I see the VHD content tests failures? I was expecting the [Test Scan and Cleanup](https://msazure.visualstudio.com/CloudNativeCompute/_build/results?buildId=102413147&view=logs&j=966f8bf1-f67e-576c-ca08-811fa43f9dc4&t=008c3696-c877-54d6-0534-28de5929aa75) step in AKS Linux VHD Build pipeline would fail for the ["always fail" test](https://github.com/Azure/AgentBaker/pull/4744/commits/19c021b4f2b2558c3adf151caa764c9b9dfcdaed) I committed, but **it's not**. @cameronmeissner

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (ganeshkumarashok) [inline parts/linux/cloud-init/artifacts/cse_install.sh:192]: Do we have plans to move wasm shims away from acs-mirror? @sozercan and @jeremyrickard mentioned that newer components and versions will need to be moved

Comment (ganeshkumarashok) [reply parts/linux/cloud-init/artifacts/cse_install.sh:192]: And also use dalec for publishing

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (ganeshkumarashok) [inline parts/linux/cloud-init/artifacts/cse_install.sh:200]: let's reduce retries here too

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (ganeshkumarashok) [inline parts/linux/cloud-init/artifacts/cse_install.sh:198]: Need to reduce retry counts here too

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (Mossaka) [reply parts/linux/cloud-init/artifacts/cse_install.sh:212]: can you please provide me some context on why we need to have such a plan?

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (Mossaka) [reply parts/linux/cloud-init/artifacts/cse_install.sh]: done.

File: pkg/agent/testdata/AKSUbuntu1604+Containerd/line40.sh
Comment (ganeshkumarashok) [inline pkg/agent/testdata/AKSUbuntu1604+Containerd/line40.sh:387]: Unnecessary whitespace change?

Comment (ganeshkumarashok) [reply pkg/agent/testdata/AKSUbuntu1604+Containerd/line40.sh:387]: I assume it's from the IDE?

File: vhdbuilder/packer/test/linux-vhd-content-test.sh
Comment (ganeshkumarashok) [inline vhdbuilder/packer/test/linux-vhd-content-test.sh:138]: Can we avoid such whitespace changes to have a cleaner diff?

File: vhdbuilder/packer/test/linux-vhd-content-test.sh
Comment (ganeshkumarashok) [inline vhdbuilder/packer/test/linux-vhd-content-test.sh:969]: Is it possible to avoid hardcoding these?

Comment (Mossaka) [reply vhdbuilder/packer/test/linux-vhd-content-test.sh:969]: well, I don't see much value from avoiding hardcoding these. I took a look at other tests and they seem to have hardcoded paths too.

**Comment by Mossaka** [comment]
@ganeshkumarashok I realized that `self-contained/bootstrap_install.sh` also defines a `downloadContainerdWasmShims()` and it's being used in `self-contained/bootstrap_main.sh`. Should I update this one as well? What's the usage of it?

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (Mossaka) [reply parts/linux/cloud-init/artifacts/cse_install.sh:212]: Thanks, I just saw the email. Ah no, we do not have plans to move wasm shims away from acs-mirror because this will be the last update we release, and after that, we will enter the deprecation process.

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (cameronmeissner) [inline parts/linux/cloud-init/artifacts/cse_install.sh:201]: is this removing support to download the shims via oras? is that something we agreed on already?

@AlisonB319

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (ganeshkumarashok) [reply parts/linux/cloud-init/artifacts/cse_install.sh:212]: Sounds good. There won't be any further changes in the agentbaker side then, right?

Comment (ganeshkumarashok) [reply parts/linux/cloud-init/artifacts/cse_install.sh:212]: If there aren't going to be any more changes in the agentbaker side afterwards, and the changes will be removed, we can be a bit more lax on parts of the reviews.

File: vhdbuilder/packer/test/linux-vhd-content-test.sh
Comment (ganeshkumarashok) [reply vhdbuilder/packer/test/linux-vhd-content-test.sh:969]: Concern is that every time we need an update or if these versions are removed from publishing or code, we need to update these test files too, or they'll fail

Comment (ganeshkumarashok) [reply vhdbuilder/packer/test/linux-vhd-content-test.sh:969]: We even had a discussion about issues related to such hardcoding today. cc: @Devinwong, @AlisonB319, are we ok with hardcoding these here? Will later refactors to consolidate to components.json and to use ORAS be affected.

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (AlisonB319) [reply parts/linux/cloud-init/artifacts/cse_install.sh:201]: Yeah we don't want to be removing oras download of wasm if these new versions can be installed during provisioning. @Mossaka is there a reason you removed this code path?

context - oras is needed to download a package when it is trying to be installed during node provisioning (aka not cached during build time).

Comment (Mossaka) [reply parts/linux/cloud-init/artifacts/cse_install.sh:201]: Ah I removed this by accident, will add it back. Good catch!

Comment (Mossaka) [reply parts/linux/cloud-init/artifacts/cse_install.sh:201]: Does `BOOTSTRAP_PROFILE_CONTAINER_REGISTRY_SERVER/oss/binaries/` have the binaries that released by the [containerd-wasm-shims pipeline](https://dev.azure.com/AzureContainerUpstream/Kubernetes/_build?definitionId=499)?

File: vhdbuilder/packer/test/linux-vhd-content-test.sh
Comment (AlisonB319) [reply vhdbuilder/packer/test/linux-vhd-content-test.sh:969]: Yeah I think it is fine to hard code for now, I'm currently working on a PR to update WASM so I'll take care of making sure they don't stay hard coded for very much longer :)

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (cameronmeissner) [reply parts/linux/cloud-init/artifacts/cse_install.sh:208]: if we're gonna support v015.1 then we should have the corresponding tag in MCR, how can we go about adding it?

Comment (AlisonB319) [reply parts/linux/cloud-init/artifacts/cse_install.sh:208]: I asked @Mossaka to reach out to Xinhe Li about adding v0.15.1. But I am unsure if 0.15.1 should block this PR since NIC is not in place yet.

Comment (cameronmeissner) [reply parts/linux/cloud-init/artifacts/cse_install.sh:208]: synced some more offline, sounds like this will be needed shortly (~2 weeks) to ensure NIC bootstrapping is stable, @Mossaka I'd prefer we get this resolved before merging this

Comment (Mossaka) [reply parts/linux/cloud-init/artifacts/cse_install.sh:208]: the MCR changes are completely out of my radar. I know @fseldow's been working on it, but not sure when it will have that in MCR.

Comment (Mossaka) [reply parts/linux/cloud-init/artifacts/cse_install.sh:208]: Would it be possible to notify me or my team in the future for any changes that made to the Wasm shims? @AlisonB319

**Comment by Mossaka** [comment]
Xinhe just informed me that the SpinKube shim v0.15.1 was pushed to MCR. I've updated the PR (in `cse_install.sh`) to reflect that change. I've also done a bit refactoring to make the code more readable. Please give it a review if you are free.

@cameronmeissner, @AlisonB319

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (Mossaka) [reply parts/linux/cloud-init/artifacts/cse_install.sh:208]: sounds good to me! Happy to review your PR.

Comment (Mossaka) [reply parts/linux/cloud-init/artifacts/cse_install.sh:208]: Gonna resolve this comment. See [my comment](https://github.com/Azure/AgentBaker/pull/4744#issuecomment-2359381086) below

File: nbcparser/pkg/parser/templates/containerd.toml.gtpl
Comment (cameronmeissner) [inline nbcparser/pkg/parser/templates/containerd.toml.gtpl:48]: could we add a new agentbakere2e validator to ensure that the containerd config is correct when using krustlet?

https://github.com/Azure/AgentBaker/blob/a9d43fb18e048d580d039de43d00eee22f1ac647/e2e/validation.go#L79

otherwise this lgtm I think

File: vhdbuilder/packer/test/linux-vhd-content-test.sh
Comment (AlisonB319) [inline vhdbuilder/packer/test/linux-vhd-content-test.sh:968]: We should add v0.8.0 v0.15.1 to cover our bases on all of the versions we are expecting

File: nbcparser/pkg/parser/templates/containerd.toml.gtpl
Comment (Mossaka) [reply nbcparser/pkg/parser/templates/containerd.toml.gtpl:48]: Added one

**Comment by Mossaka** [comment]
It looks like all the tests, including the e2e Wasm validators, were passing, except 1804 builds. I've update the self-containerd to have the mirrored logic from cse_install.

**Review by cameronmeissner — APPROVED** [review]

**Pull request merged**
